File: java/Trees/Serialize_and_Deserialize_Binary_Tree.java
package java.Trees;

import java.util.*;

class TreeNode {
  int val;
  TreeNode left, right;

  TreeNode(int val) {
    this.val = val;
  }
}

public class Serialize_and_Deserialize_Binary_Tree {
  public String serialize(TreeNode root) {
    if (root == null)
      return "null";
    return root.val + "," + serialize(root.left) + "," + serialize(root.right);
  }

  public TreeNode deserialize(String data) {
    Queue<String> queue = new LinkedList<>(Arrays.asList(data.split(",")));
    return build(queue);
  }

  TreeNode build(Queue<String> queue) {
    String val = queue.poll();
    if (val.equals("null"))
      return null;
    TreeNode node = new TreeNode(Integer.parseInt(val));
    node.left = build(queue);
    node.right = build(queue);
    return node;
  }
}

File: C++/Tree/Serialize_and_Deserialize_Binary_Tree.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for a binary tree node.
struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Codec
{
public:
  string serialize(TreeNode *root)
  {
    if (!root)
      return "#,";

    return to_string(root->val) + "," + serialize(root->left) + serialize(root->right);
  }

  TreeNode *deserialize(string data)
  {
    stringstream ss(data);
    return build(ss);
  }

private:
  TreeNode *build(stringstream &ss)
  {
    string val;
    getline(ss, val, ',');

    if (val == "#")
      return nullptr;

    TreeNode *root = new TreeNode(stoi(val));

    root->left = build(ss);
    root->right = build(ss);

    return root;
  }
};